File: c_files/src/Planet.c
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.05.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/

#include "Planet.h"
#include "SpaceShip.h"
#include "StringUtils.h"
#include "MyTime.h"

/* PRIVATE FUNCTIONS */
void	updateCurrentHour(Planet this)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL)
	{
		printf("Error: updateCurrentHour 'this' is NULL!\n");
		return;
	}

	/* TIME STRUCT FOR GLOBAL STATIC TIME */
	MyTime time = new_Time();
	if (time == NULL)
	{
		printf("Memory allocation failed at updateCurrentHour\n");
		return;
	}

	/* UPDATE CURRENT HOUR*/
	this->currentHour = (int)(time->getGlobalTime()% this->hoursOfADay);
	if (this->currentHour == 0)
	{
		this->currentDate->advanceDay(this->currentDate, 1);
	}

	/* DELETE TIME STRUCT */
	time->delete_Time(time);
}
/* END OF PRIVATE FUNCTIONS */

Planet	new_Planet(const char* name, int planetType, int hoursOfADay, int day, int month, int year)
{
	/* CHECKING THE PARAMETERS */
	if (name == NULL)
	{
		printf("Memory allocation failed at new_Planet\n");
		return (NULL);
	}

	/* MEMORY ALLOCATION FOR PLANET*/
	Planet this = (Planet)malloc(sizeof(struct PLANET));
	if (this == NULL)
	{
		printf("Memory allocation failed at new_Planet\n");
		return (NULL);
	}
	
	/* MEMORY ALLOCATION FOR TIMELISTENER */
	this->listener = new_ITimeListener();
	if (this->listener == NULL)
	{
		printf("Memory allocation failed for listener!\n");
		return (NULL);
	}

	/* MEMORY ALLOCATION FOR MYLOCALTIME*/
	this->currentDate = new_MyLocalTime(day, month, year);
	if (this->currentDate == NULL)
	{
		printf("Memory allocation failed at new_Planet MyLocalTime\n");
		free(this->listener);
		free(this);
		return (NULL);
	}

	/* SET INITIAL VALUES */
	this->name = strdup(name);
	this->planetType = planetType;
	this->hoursOfADay = hoursOfADay; 
	this->currentHour = 0;
	this->population = 0;

	/* LINKS */
	this->getDecayRate = &getDecayRate;
	this->setInitials_Planet = &setInitials_Planet;
	this->updatePopulation = &updatePopulation;
	this->delete_Planet = &delete_Planet;
	this->listener->onTimePass = &onTimePass_Planet; // interface implementation method
	/* END OF LINKS */

	return (this);
}

void	setInitials_Planet(const Planet this, ArrayList shipList)
{
	/* CHECKING THE PARAMETERS*/
    if (this == NULL || shipList == NULL) {
        printf("EROR: at setInitials_Planet\n");
        return;
    }
	/* SET */
    this->allShips = shipList;
}

double	getDecayRate(const Planet this)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL)
	{
		printf("Error: at getDecayRate\n");
		return 0.0;
	}

	/* DETERMINE TYPE */
	switch (this->planetType)
	{
		case 0: // Terrestrial
			return 1.0;
		case 1: // Gas Giant
			return 0.1;
		case 2: // Ice Giant
			return 0.5;
		case 3: // Dwarf Planet
			return 0.01;
		default:
			return 1.0;
	}
}

void	updatePopulation(const Planet this)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL || this->allShips == NULL)
	{
		printf("Hata: updatePopulation içinde 'this' NULL!\n");
		return;
	}

	/* CALCULATE POPULATION */
	int population = 0;
	for (size_t i = 0; i < this->allShips->size; i++)
	{
		SpaceShip ship = (SpaceShip)(this->allShips->getByIndex(this->allShips, i));
		/* CHECKING THE VARIABLES */
		if (ship != NULL || 
			ship->currentPlanetName != NULL ||
			ship->destPlanetName != NULL ||
			ship->state != NULL)
		{
			if (
				((equalsIgnoreCase(ship->currentPlanetName, this->name)) && (strcmp(ship->state, "Bekliyor") == 0)) ||
				((equalsIgnoreCase(ship->destPlanetName, this->name)) && (strcmp(ship->state, "Vardi") == 0))
			)
			{ // IF CONDITIONS ARE MET COUNT PASSANGER AS POPULATION
				population += ship->passengersCount;
			}
		}
	}
	/* SET VALUE */
	this->population = population;
}

void	delete_Planet(Planet this)
{
	if (this == NULL)
		return;
    
	if (this->name != NULL)
		free(this->name);

	if (this->listener != NULL)
		free(this->listener);

	if (this->currentDate != NULL)
		this->currentDate->delete_MyLocalTime(this->currentDate);

	free(this);
}

/* OVERRIDE METHOD */
void	onTimePass_Planet(const Planet this)
{
	if (this == NULL)
	{
		printf("Error: onTimePass_Planet !\n");
		return;
	}

	updateCurrentHour(this);
	updatePopulation(this);
}

File: java_files/pdp_planet_travel_sim/src/app/Planet.java
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.04.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/

package app;

/* IMPORTS */
import java.util.List;
/*END OF IMPORTS */

public class Planet implements ITimeListener
{
	private String			name;		// TO SET NAME THAT WE GET FROM THE FILE
	private int				hoursOfADay;// TO SET THE HOURS OF A DAY THAT WE GET FROM THE FILE
	private MyLocalTime		currentDate;// TO SET THE CURRENT DATE THAT WE GET FROM THE FILE
	private int				currentHour;// TO SET THE CURRENT HOUR THAT WE GET FROM THE FILE
	private int				population;	// TO SET THE POPULATION OF THE PLANET
	private List<SpaceShip>	ships;		// TO CALCULATE HOW MANY PEOPLE IS CURRENTLY LOCATED IN THE PLANET
	
	/* CONSTRUCTER */
	
	public Planet(String name, int hoursOfADay, MyLocalTime currentDate)
	{
		setName(name);
		setHoursOfADay(hoursOfADay);
		setCurrentDate(currentDate);
		setCurrentHour();
		setPopulation();
		Time.addListener(this);
	}
	
	/*END OF CONSTRUCTER*/
	
	/* SETTER METHODS */
	
	private void	setName(String name)
	{
		this.name = name;
	}
	private void	setHoursOfADay(int hour)
	{
		this.hoursOfADay = hour;
	}
	private void	setCurrentDate(MyLocalTime date)
	{
		this.currentDate = date;
	}
	private void	setCurrentHour()
	{
		this.currentHour = 0;
	}
	private void	setPopulation()
	{
		this.population = 0;
	}
	public void		setSpaceShipList(List<SpaceShip> ships)
	{
		this.ships = ships;
	}
	
	/* END OF SETTER METHODS */

	/* GETTER METHODS */
	
	public String		getName()
	{
		return (name);
	}
	public MyLocalTime	getCurrentDate()
	{
		return (currentDate);
		
	}
	public Integer 		getHoursOfADay()
	{
		return (hoursOfADay);
	}
	public int	 		getPopulation()
	{
		return (population);
	}
	
	
	/* END OF GETTER METHODS */
	
	/* METHODS */
	
	private void	updateCurrentHour()
	{
		// CALCULATE CURRENT HOUR
		this.currentHour = (int) (Time.getGlobalTime() % this.hoursOfADay);
		
		if (this.currentHour == 0)
		{
			// ADVANCE THE A DAY IF THE DAY IS ENDED
			this.currentDate.advanceDay();
		}
	}
	
	public void		updatePopulation()
	{	
		// START SHIP STREAM
		// IF CURRENT PLANET NAME AND THIS.NAME ARE MATCHED AND THE STATE OF THE SHIP IS "BEKLİYOR"
		// OR
		// IF DESTINATION PLANET NAME AND THIS.NAME ARE MATCHED AND THE STATE OF THE SHIP IS "VARDI"
		// CONVERT AN OBJECT STREAM TO AN INT STREAM 
		// SUM TO ADD THE POPULATION
		this.population = this.ships.stream()
						.filter(s -> (s.getCurrentPlanetName().equalsIgnoreCase(this.name) && s.getState().equals("Bekliyor"))
						|| (s.getDestPlanetName().equalsIgnoreCase(this.name) && s.getState().equals("Vardı")) )
						.mapToInt(SpaceShip :: getPassangersCount)
						.sum();
	}
	

	
	@Override
	public void		onTimePass() 
	{
		/* TRIGGER THE FUNCTIONS THAT RELATED TO THE TIME */
		updateCurrentHour();
		updatePopulation();

	}
}
